File: backend/src/main/java/com/flowlink/transform/TransformRegistry.java
package com.flowlink.transform;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.net.URL;
import java.net.URLClassLoader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.ServiceLoader;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Stream;

/**
 * 변환 레지스트리 — 내장 변환 + 플러그인 디렉토리의 JAR(ServiceLoader)을 보유한다.
 * (신뢰 JAR 전용 — 샌드박스 없음. 관리자만 업로드)
 */
@Component
public class TransformRegistry {

    private static final Logger log = LoggerFactory.getLogger(TransformRegistry.class);

    private final Map<String, FlowTransform> byId = new ConcurrentHashMap<>();
    private final Path pluginDir;

    public TransformRegistry(@Value("${flowlink.plugins.dir:plugins}") String dir) {
        this.pluginDir = Path.of(dir);
        reload();
    }

    /** 내장 + 플러그인을 다시 스캔해 등록. (JAR 업로드 후 호출) */
    public final synchronized void reload() {
        Map<String, FlowTransform> next = new LinkedHashMap<>();
        for (FlowTransform t : BuiltinTransforms.all()) {
            next.put(t.id(), t);
        }
        int pluginCount = loadPlugins(next);
        byId.clear();
        byId.putAll(next);
        log.info("변환 레지스트리 로드: 내장 {}, 플러그인 {} (총 {})",
                BuiltinTransforms.all().size(), pluginCount, byId.size());
    }

    private int loadPlugins(Map<String, FlowTransform> map) {
        if (!Files.isDirectory(pluginDir)) {
            return 0;
        }
        int count = 0;
        try (Stream<Path> stream = Files.list(pluginDir)) {
            List<URL> jars = new ArrayList<>();
            for (Path p : stream.filter(p -> p.toString().toLowerCase().endsWith(".jar")).toList()) {
                jars.add(p.toUri().toURL());
            }
            if (jars.isEmpty()) {
                return 0;
            }
            URLClassLoader cl = new URLClassLoader(jars.toArray(URL[]::new), getClass().getClassLoader());
            for (FlowTransform t : ServiceLoader.load(FlowTransform.class, cl)) {
                map.put(t.id(), t); // 플러그인이 내장을 덮어쓸 수 있음
                count++;
            }
        } catch (Exception e) {
            log.warn("플러그인 로드 실패: {}", e.getMessage());
        }
        return count;
    }

    public Optional<FlowTransform> get(String id) {
        return Optional.ofNullable(byId.get(id));
    }

    public List<FlowTransform> list() {
        return new ArrayList<>(byId.values());
    }
}

File: backend/src/main/kotlin/com/flowlink/transform/TransformRegistry.kt
package com.flowlink.transform

import com.flowlink.codec.CodecPlugin
import com.flowlink.plugin.PluginsProperties
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.net.URLClassLoader
import java.nio.file.Files
import java.nio.file.Path
import java.util.Optional
import java.util.ServiceLoader
import java.util.concurrent.ConcurrentHashMap

/** 승인된 스크립트 플러그인을 SPI 구현체(FlowTransform | CodecPlugin)로 넘겨주는 소스 — DB 로더가 구현(PluginScriptService). */
fun interface ScriptPluginLoader {
    fun loadApproved(): List<Any>
    companion object { @JvmField val NONE = ScriptPluginLoader { emptyList() } }
}

/**
 * 변환/코덱 레지스트리 — 스크립트 플러그인(DB 승인본) + (jar-enabled 일 때만) 플러그인 디렉터리의 JAR(ServiceLoader).
 * JAR 은 신뢰 전제·샌드박스 없음이라 기본 비활성. 같은 id 는 나중 로드가 덮는다(스크립트가 JAR 뒤에 로드 → 스크립트 우선).
 */
@Component
class TransformRegistry(
    private val props: PluginsProperties,
    private val scripts: ScriptPluginLoader = ScriptPluginLoader.NONE,
) {
    private val byId: MutableMap<String, FlowTransform> = ConcurrentHashMap()
    private val codecById: MutableMap<String, CodecPlugin> = ConcurrentHashMap()
    private val pluginDir: Path = Path.of(props.dir)
    @Volatile private var loader: URLClassLoader? = null

    init { reload() }

    /** 다시 스캔해 등록(스크립트 승인/삭제·JAR 배치 후). 이전 URLClassLoader 는 닫는다(Windows JAR 잠김 방지). */
    @Synchronized
    fun reload() {
        val next = LinkedHashMap<String, FlowTransform>()
        val nextCodecs = LinkedHashMap<String, CodecPlugin>()
        val jars = if (props.jarEnabled) loadJars(next, nextCodecs) else warnSkippedJars()
        var scriptCount = 0
        for (p in scripts.loadApproved()) {
            when (p) {
                is FlowTransform -> { next[p.id()] = p; scriptCount++ }
                is CodecPlugin -> { nextCodecs[p.id()] = p; scriptCount++ }
            }
        }
        byId.clear(); byId.putAll(next)
        codecById.clear(); codecById.putAll(nextCodecs)
        log.info("플러그인 로드 — JAR {}개, 스크립트 {}개 → 변환 {}개, 코덱 {}개", jars, scriptCount, byId.size, codecById.size)
    }

    private fun jarFiles(): List<Path> =
        if (!Files.isDirectory(pluginDir)) emptyList()
        else Files.list(pluginDir).use { s -> s.filter { it.toString().lowercase().endsWith(".jar") }.sorted().toList() }

    private fun warnSkippedJars(): Int {
        val jars = jarFiles()
        if (jars.isNotEmpty()) log.warn("플러그인 JAR {}개가 있으나 flowlink.plugins.jar-enabled=false — 로드하지 않음: {}", jars.size, jars.map { it.fileName.toString() })
        return 0
    }

    /** JAR 하나씩 별도 로더로 — 깨진 JAR 이 나머지를 못 죽인다. 반환: 성공한 JAR 수. */
    private fun loadJars(map: MutableMap<String, FlowTransform>, codecs: MutableMap<String, CodecPlugin>): Int {
        loader?.let { runCatching { it.close() } }
        val jars = jarFiles()
        if (jars.isEmpty()) return 0
        var ok = 0
        val cl = URLClassLoader(jars.map { it.toUri().toURL() }.toTypedArray(), javaClass.classLoader)
        loader = cl
        for (jar in jars) {
            try {
                val one = URLClassLoader(arrayOf(jar.toUri().toURL()), cl)
                for (t in ServiceLoader.load(FlowTransform::class.java, one)) map[t.id()] = t
                for (c in ServiceLoader.load(CodecPlugin::class.java, one)) codecs[c.id()] = c
                ok++
            } catch (e: Throwable) { // ServiceConfigurationError 는 Error 계열
                log.warn("플러그인 JAR 로드 실패(건너뜀): {} — {}", jar.fileName, e.message ?: e.toString())
            }
        }
        return ok
    }

    fun get(id: String): Optional<FlowTransform> = Optional.ofNullable(byId[id])
    fun list(): List<FlowTransform> = ArrayList(byId.values)
    fun codec(id: String): CodecPlugin? = codecById[id]
    fun codecs(): List<CodecPlugin> = ArrayList(codecById.values)

    companion object { private val log = LoggerFactory.getLogger(TransformRegistry::class.java) }
}
